include/rdma/fabric.h

```diff
@@ -347,6 +370,7 @@ struct fi_rx_attr {
 	size_t			total_buffered_recv;
 	size_t			size;
 	size_t			iov_limit;
+	struct fi_traffic_class traffic_class;
```

Comment:
I had that in my original pull request for internal review, but an argument was raised for having them on a TX/RX basis. The primary argument for having them split apart was support for finer grained  control of traffic class for multi-transaction data transfer operations like rendezvous. It was suggested that it would be nice if an application could specify what traffic class that a rendezvous response would be sent on. 

For example:
As an application writer, I want my outbound traffic to go out on a specific traffic class, but if my request would result in a rendezvous message where the payload was read from me, or sent to me from the peer, I want that traffic to go out on a different traffic class. In practice, the TX traffic class might FI_TC_DEDICATED_ACCESS, and the RX/rendezvous traffic class would be FI_TC_BULK_IPC_TRAFFIC. 

From the commit above: 
```
The original proposal to OFIWG did not include the rx context modification. However,
multi-part data transfer operations such as rendezvous messaging benefit from the differentiation of
request and response traffic shaping flows. The traffic class field in the rx attribute is
only intended to apply to these multi-part data transfer operations and does not imply any
relation with the tx attribute of a peer endpoint.

```

Reply:
Traffic classes apply to sent data, not receives.  Trying to assign a traffic class to a receive queue doesn't make sense.

I saw your comment.  But having the rx and tx be separate classes is odd.  It's more than rendezvous, which generate tx traffic anyway.  Are we saying acks are sent using a separate tclass than that used by the sender?  What does that even mean in terms of the actual class then?

If we really want rendezvous to use a separate class, then we're looking at assigning a tclass per transfer.  We agreed in the ofiwg that that was overkill.

Reply:
I get what you mean. I had this discussion internally here, and the result of that discussion was a desire to provide a different traffic class to rendezvous responses which may carry a much larger payload than the initial rendezvous request. 

> Are we saying acks are sent using a separate tclass than that used by the sender?

I think that's up to us, or up to the providers to define. 

>  What does that even mean in terms of the actual class then?

As far as I'm aware, this means that a rendezvous request uses TC A to make its way to the peer, and the rendezvous response uses TC B to fetch the data from the originator. -- Just as an example. 

> If we really want rendezvous to use a separate class, then we're looking at assigning a tclass per transfer.

I'm avoiding that as well. I think providing an RX TC allows enough control over multi-transaction data transfers to do what it needs to do without specifying traffic class on a per-message basis.

Reply:
Ignore rendezvous.  That's an implementation issue for performing large transfers.  Assigning a tclass to the receive queue makes no sense from the viewpoint of an API.  By definition, a receive queue doesn't transmit.

The only reasonable interpretation of assigning a tclass to the rx is that it affects packets generated (likely by hardware) in response to a received message.  That includes ACKs, atomic responses, RMA read responses, some triggered operations, and collective operations.

I question the need to support these packets using a different tclass than packets resulting from application issued transmits.

These are all separate from rendezvous.  Rendezvous is (usually) a transmit initiated by software in response to a received message.  Rendezvous can be implemented using RMA reads or writes or even tagged messages, among other options.  This is basically asking for a tclass per transmit.  That meets the needs of supporting rendezvous generated RMA at a different tclass than, say, local application issued RMA.

Adding an rx tclass because of how a provider has implemented large message transfers isn't the right approach.  If this level of granularity is needed, we should move to having per transmit controls on the tclass.  I expect that would be unusable in practice however.

Reply:
> Assigning a tclass to the receive queue makes no sense from the viewpoint of an API. By definition, a receive queue doesn't transmit.

I agree completely. 

> Adding an rx tclass because of how a provider has implemented large message transfers isn't the right approach.

I agree. Actually after some thought, I think that if a provider wants to implement this type of behavior, they can use the existing provider-specific mechanisms (provider-specific get/setopt functions) to allow applications to set this type of behavior -- rather than trying to stand it up in the API. 

------------

I'm going to yank this out. I didn't like it but it was an effort on my part to present the idea anyway. Thanks.